From 2d16c73564183b3698d9d59a68a8b79695207475 Mon Sep 17 00:00:00 2001
From: Ding Wei <leo.ding@rock-chips.com>
Date: Fri, 12 Jun 2020 17:01:47 +0800
Subject: [PATCH] video: rockchip: mpp: refresh pm_runtime for shared iommu

when iommu shared, such as px30 rk3368, each device power on affect
iommu usage_count add once. in this case, it should pm_runtime_put/get
by usage_count times,then hardware will really suspend/resume.

Change-Id: I7e57818e4eaf8af84e52c2d77030791329721c2e
Signed-off-by: Ding Wei <leo.ding@rock-chips.com>
---
 drivers/video/rockchip/mpp/mpp_common.c | 21 +++++++++++++++------
 1 file changed, 15 insertions(+), 6 deletions(-)

diff --git a/drivers/video/rockchip/mpp/mpp_common.c b/drivers/video/rockchip/mpp/mpp_common.c
index d63a5404f502..dd4f8faac540 100644
--- a/drivers/video/rockchip/mpp/mpp_common.c
+++ b/drivers/video/rockchip/mpp/mpp_common.c
@@ -368,17 +368,26 @@ static int mpp_process_task(struct mpp_session *session,
 	return 0;
 }
 
-static int mpp_refresh_pm_runtime(struct device *dev)
+static int mpp_refresh_pm_runtime(struct mpp_iommu_info *info,
+				  struct device *dev)
 {
+	int i;
+	int usage_count;
 	struct device_link *link;
+	struct device *iommu_dev = &info->pdev->dev;
 
 	rcu_read_lock();
 
-	list_for_each_entry_rcu(link, &dev->links.suppliers, c_node)
-		pm_runtime_put_sync(link->supplier);
+	usage_count = atomic_read(&iommu_dev->power.usage_count);
+	list_for_each_entry_rcu(link, &dev->links.suppliers, c_node) {
+		for (i = 0; i < usage_count; i++)
+			pm_runtime_put_sync(link->supplier);
+	}
 
-	list_for_each_entry_rcu(link, &dev->links.suppliers, c_node)
-		pm_runtime_get_sync(link->supplier);
+	list_for_each_entry_rcu(link, &dev->links.suppliers, c_node) {
+		for (i = 0; i < usage_count; i++)
+			pm_runtime_get_sync(link->supplier);
+	}
 
 	rcu_read_unlock();
 
@@ -495,7 +504,7 @@ int mpp_dev_reset(struct mpp_dev *mpp)
 	 * as an empty operation. Therefore, force to close and then open,
 	 * will be update the domain. In this way, domain can really attach.
 	 */
-	mpp_refresh_pm_runtime(mpp->dev);
+	mpp_refresh_pm_runtime(mpp->iommu_info, mpp->dev);
 
 	mpp_iommu_attach(mpp->iommu_info);
 	up_write(&mpp->reset_group->rw_sem);
-- 
2.35.3

